samza-core/src/main/java/org/apache/samza/zk/ZkBarrierForVersionUpgrade.java

```diff
@@ -117,41 +132,50 @@ public void join(String version, String participantId) {
    * @param version Version associated with the Barrier
    */
   public void expire(String version) {
-    zkUtils.writeData(keyBuilder.getBarrierStatePath(version), State.TIMED_OUT);
-
+    String barrierStatePath = keyBuilder.getBarrierStatePath(version);
+    State barrierState = zkUtils.getZkClient().readData(barrierStatePath);
+    if (Objects.equals(barrierState, State.NEW)) {
+      LOG.info(String.format("Expiring the barrier version: %s. Marking the barrier state: %s as %s.", version, barrierStatePath, State.TIMED_OUT));
+      zkUtils.writeData(keyBuilder.getBarrierStatePath(version), State.TIMED_OUT);
```

Comment:
Not necessarily in this PR: For debugability, it will be helpful to write strings to zookeeper instead of objects. Often, open-source tools like _zooinspector_ and internal UIs like _reflection_ work better with UTF-8 encoded strings.